exchanges/orderbook/linked_list.go

```diff
@@ -0,0 +1,675 @@
+		var bookmark *node
+		for tip := ll.head; tip != nil; tip = tip.next {
+			if tip.value.ID == updts[x].ID {
+				if tip.value.Price != updts[x].Price { // Price level change
+					// bookmark tip to move this node to correct price level
+					bookmark = tip
+					continue // continue through node depth
+				}
+				// no price change, amend amount and conintue updates
+				tip.value.Amount = updts[x].Amount
+				continue updates // continue to next update
+			}
+
+			if tip.value.Price > updts[x].Price {
+				if bookmark != nil { // shift bookmarked node to current tip
+					bookmark.value = updts[x]
+					move(&ll.head, bookmark, tip)
+					continue updates
+				}
+
+				// search for ID
+				for n := tip.next; n != nil; n = n.next {
+					if n.value.ID == updts[x].ID {
+						n.value = updts[x]
+						// inserting before the tip
+						move(&ll.head, n, tip)
+						continue updates
+					}
+				}
+				// ID not matched in depth so add correct level for insert
+				bookmark = tip
+				break
+			}
+
+			if tip.next == nil {
+				if bookmark == nil {
+					bookmark = tip
+				} else {
+					bookmark.value = updts[x]
+					bookmark.next.prev = bookmark.prev
+					if bookmark.prev == nil { // Bookmark is at head
+						ll.head = bookmark.next
+					} else {
+						bookmark.prev.next = bookmark.next
+					}
+					tip.next = bookmark
+					bookmark.prev = tip
+					bookmark.next = nil
+					continue updates
+				}
+			}
+		}
+		n := stack.Pop()
+		n.value = updts[x]
+		switch {
+		case bookmark.prev == nil:
+			ll.head = n
+			n.prev = bookmark.prev
+			bookmark.prev = n
+			n.next = bookmark
+		case bookmark.next == nil:
+			n.prev = bookmark
+			bookmark.next = n
+		default:
+			bookmark.prev.next = n
+			n.prev = bookmark.prev
+			bookmark.prev = n
+			n.next = bookmark
+		}
+		ll.length++
+	}
+	return nil
+}
+
+// insertUpdates inserts new updates for asks based on price level
+func (ll *asks) insertUpdates(updts Items, stack *stack) error {
+	for x := range updts {
+		var prev *node
+		for tip := &ll.head; ; tip = &(*tip).next {
+			if *tip == nil { // Head is empty
+				// This is here because there might be an instance where the
+				// liquidity on an exchange does fall to zero through a
+				// streaming endpoint then it comes back online.
+				n := stack.Pop()
+				n.value = updts[x]
+				n.prev = prev
+				ll.length++
+				*tip = n
+				break // Continue updates
+			}
+
+			if (*tip).value.Price == updts[x].Price { // Price already found
+				return fmt.Errorf("%w for price %f",
+					errCollisionDetected,
+					updts[x].Price)
+			}
+
+			// Correct position/allignment found for price level
+			if (*tip).value.Price > updts[x].Price {
+				n := stack.Pop()
+				n.value = updts[x]
+				n.prev = prev
+				ll.length++
+				if (*tip).prev == nil {
+					// Place new node in front of current node
+					(*tip).prev = n
+					n.next = *tip
+					// Replace head entry
+					*tip = n
+				} else {
+					old := *tip
+					prev.next = n
```

Comment:
prev hasn't been checked here and could cause a panic

Reply:
current (*tip).prev == prev; prev.next should == *tip. 
I think construction wise there has been a serious fault and it should panic because all subsequant nodes are off. Have you been able to replicate a panic?

Reply:
I don't understand the way you wrote your example code sorry, I'll speak to you soon to get a better understanding. prev doesn't get assigned to until the end of the loop. My IDE is flagging it, it will take me longer to understand everything here, but I haven't yet seen a reason not to protect it. It seems like a scenario can exist where this can be hit and prev has either been assigned nil or it hasn't been set yet.

Leaning on a panic isn't something we should be doing. Better to be able to identify the issue and handle it instead.